src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs

```diff
@@ -11,27 +12,36 @@ namespace Microsoft.TemplateEngine.Cli.Commands
 {
     internal partial class InstantiateCommand : BaseCommand<InstantiateCommandArgs>
     {
-        internal IEnumerable<string> GetSuggestions(
+        internal IEnumerable<CompletionItem> GetCompletions(
             InstantiateCommandArgs args,
             IEnumerable<TemplateGroup> templateGroups,
             IEngineEnvironmentSettings environmentSettings,
             TemplatePackageManager templatePackageManager,
-            string? textToMatch)
+            TextCompletionContext context)
         {
             if (string.IsNullOrWhiteSpace(args.ShortName))
             {
-                return templateGroups.SelectMany(g => g.ShortNames).Distinct().OrderBy(s => s, StringComparer.OrdinalIgnoreCase);
+                return templateGroups
+                    .SelectMany(g => g.ShortNames, (g, shortName) => new CompletionItem(shortName, documentation: g.Description))
+                    .Distinct()
+                    .OrderBy(c => c.Label, StringComparer.OrdinalIgnoreCase)
+                    .Concat(base.GetCompletions(context, environmentSettings));
             }
             else
             {
                 //no exact match on short name
                 if (!templateGroups.Any(template => template.ShortNames.Contains(args.ShortName)))
                 {
-                    return templateGroups.SelectMany(g => g.ShortNames).Where(n => n.StartsWith(args.ShortName)).Distinct().OrderBy(s => s, StringComparer.OrdinalIgnoreCase);
+                    return templateGroups
+                        .SelectMany(g => g.ShortNames, (g, shortName) => new CompletionItem(shortName, documentation: g.Description))
+                        .Where(c => c.Label.StartsWith(args.ShortName))
+                        .Distinct()
+                        .OrderBy(c => c.Label, StringComparer.OrdinalIgnoreCase)
+                        .Concat(base.GetCompletions(context, environmentSettings));
                 }
 
                 //if there is exact match do further reparsing
-                HashSet<string> distinctSuggestions = new HashSet<string>();
+                HashSet<CompletionItem> distinctCompletions = new HashSet<CompletionItem>();
```

Comment:
Same concern with multiple `CompletionItem`s containing the same shortName.

Reply:
same resolution as above.

Reply:
Can we remove `Distinct` calls then, since they are not doing anything? This HashSet as well.

Reply:
You can ignore my last comment here. More explanation on comment: https://github.com/dotnet/templating/pull/4193#discussion_r769522710